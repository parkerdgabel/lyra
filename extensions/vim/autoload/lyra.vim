" Vim autoload functions for Lyra
" Language: Lyra
" Maintainer: Lyra Language Team

" Completion function for Lyra
function! lyra#complete(findstart, base)
  if a:findstart
    " Find the start of the current word
    let line = getline('.')
    let start = col('.') - 1
    while start > 0 && line[start - 1] =~ '\a\|_\|\d'
      let start -= 1
    endwhile
    return start
  else
    " Return completion matches
    let completions = []
    
    " Add built-in functions
    let builtin_functions = [
          \ 'Sin', 'Cos', 'Tan', 'ArcSin', 'ArcCos', 'ArcTan',
          \ 'Sinh', 'Cosh', 'Tanh', 'Log', 'Ln', 'Exp', 'Sqrt',
          \ 'Abs', 'Floor', 'Ceil', 'Round', 'Max', 'Min',
          \ 'Sum', 'Product', 'Mean', 'Median', 'Variance',
          \ 'Length', 'First', 'Last', 'Rest', 'Most', 'Take', 'Drop',
          \ 'Join', 'Sort', 'Reverse', 'Union', 'Intersection',
          \ 'Map', 'Apply', 'Select', 'Cases', 'Count', 'Fold',
          \ 'MatchQ', 'Replace', 'ReplaceAll', 'Position',
          \ 'If', 'Which', 'Switch', 'While', 'For', 'Do',
          \ 'Module', 'Block', 'With', 'Function',
          \ 'Print', 'Echo', 'Input', 'Get', 'Put'
          \ ]
    
    " Add mathematical constants
    let math_constants = [
          \ 'Pi', 'E', 'I', 'Infinity', 'ComplexInfinity',
          \ 'GoldenRatio', 'EulerGamma', 'Catalan'
          \ ]
    
    " Add Greek letters
    let greek_letters = [
          \ 'Alpha', 'Beta', 'Gamma', 'Delta', 'Epsilon', 'Zeta',
          \ 'Eta', 'Theta', 'Iota', 'Kappa', 'Lambda', 'Mu',
          \ 'Nu', 'Xi', 'Omicron', 'Rho', 'Sigma', 'Tau',
          \ 'Upsilon', 'Phi', 'Chi', 'Psi', 'Omega'
          \ ]
    
    " Filter completions based on input
    for item in builtin_functions + math_constants + greek_letters
      if item =~? '^' . a:base
        call add(completions, {
              \ 'word': item,
              \ 'menu': '[Lyra]',
              \ 'kind': item =~? '^\(Sin\|Cos\|Tan\|Log\|Exp\|Sqrt\|Abs\|Floor\|Ceil\|Round\|Max\|Min\)' ? 'f' : 
              \         item =~? '^\(Pi\|E\|I\|Infinity\|Alpha\|Beta\|Gamma\)' ? 'v' : 'f'
              \ })
      endif
    endfor
    
    " Add variables from current buffer
    let variables = lyra#get_buffer_variables()
    for var in variables
      if var =~? '^' . a:base
        call add(completions, {
              \ 'word': var,
              \ 'menu': '[var]',
              \ 'kind': 'v'
              \ })
      endif
    endfor
    
    return completions
  endif
endfunction

" Get variables defined in current buffer
function! lyra#get_buffer_variables()
  let variables = []
  let lines = getline(1, '$')
  
  for line in lines
    " Match variable assignments: varName = value or varName := value
    let matches = matchlist(line, '\([a-zA-Z][a-zA-Z0-9]*\)\s*:*=')
    if !empty(matches)
      let var = matches[1]
      if index(variables, var) == -1
        call add(variables, var)
      endif
    endif
    
    " Match function definitions: funcName[args_] := body
    let matches = matchlist(line, '\([a-zA-Z][a-zA-Z0-9]*\)\[.*\]\s*:*=')
    if !empty(matches)
      let func = matches[1]
      if index(variables, func) == -1
        call add(variables, func)
      endif
    endif
  endfor
  
  return variables
endfunction

" Format Lyra code
function! lyra#format()
  " Save cursor position
  let save_cursor = getcurpos()
  
  " Format the entire buffer
  silent! execute '%s/\s\+$//e'  " Remove trailing whitespace
  silent! execute '%s/\(\S\)\s*\([=!<>]=\?\|&&\|||\)/\1 \2/ge'  " Add spaces around operators
  silent! execute '%s/\([=!<>]=\?\|&&\|||\)\s*\(\S\)/\1 \2/ge'  " Add spaces around operators
  silent! execute '%s/,\s*/, /ge'  " Normalize comma spacing
  silent! execute '%s/{\s*/, /ge'  " Normalize brace spacing
  silent! execute '%s/\s*}/}/ge'  " Normalize brace spacing
  
  " Restore cursor position
  call setpos('.', save_cursor)
  
  echo "Lyra code formatted"
endfunction

" Lint Lyra code (basic syntax checking)
function! lyra#lint()
  let errors = []
  let lines = getline(1, '$')
  let line_num = 0
  
  for line in lines
    let line_num += 1
    
    " Check for unbalanced brackets
    let brackets = {'[': 0, '{': 0, '(': 0}
    let i = 0
    while i < len(line)
      let char = line[i]
      if has_key(brackets, char)
        let brackets[char] += 1
      elseif char == ']'
        let brackets['['] -= 1
      elseif char == '}'
        let brackets['{'] -= 1
      elseif char == ')'
        let brackets['('] -= 1
      endif
      let i += 1
    endwhile
    
    for [bracket, count] in items(brackets)
      if count != 0
        call add(errors, {
              \ 'lnum': line_num,
              \ 'col': 1,
              \ 'text': 'Unbalanced ' . bracket . ' bracket',
              \ 'type': 'E'
              \ })
      endif
    endfor
    
    " Check for invalid operators
    if line =~ '\(===\|&&&\||||\)'
      call add(errors, {
            \ 'lnum': line_num,
            \ 'col': match(line, '\(===\|&&&\||||\)') + 1,
            \ 'text': 'Invalid operator',
            \ 'type': 'E'
            \ })
    endif
    
    " Check for unterminated strings
    if line =~ '"\([^"]\|\\.\)*$\|''\([^'']\|\\.\)*$'
      call add(errors, {
            \ 'lnum': line_num,
            \ 'col': match(line, '"\([^"]\|\\.\)*$\|''\([^'']\|\\.\)*$') + 1,
            \ 'text': 'Unterminated string',
            \ 'type': 'E'
            \ })
    endif
  endfor
  
  " Set quickfix list
  call setqflist(errors, 'r')
  
  if len(errors) > 0
    echo 'Found ' . len(errors) . ' errors. Use :copen to view them.'
    copen
  else
    echo 'No syntax errors found.'
    cclose
  endif
endfunction

" Export buffer to different formats
function! lyra#export(format)
  if &filetype != 'lyra'
    echo "Current buffer is not a Lyra file"
    return
  endif
  
  let filename = expand('%')
  if filename == ''
    echo "Buffer has no associated file. Please save first."
    return
  endif
  
  " Save file if modified
  if &modified
    write
  endif
  
  let base_name = fnamemodify(filename, ':r')
  
  if a:format == 'jupyter'
    let output_file = base_name . '.ipynb'
  elseif a:format == 'latex'
    let output_file = base_name . '.tex'
  elseif a:format == 'html'
    let output_file = base_name . '.html'
  else
    echo "Unknown format: " . a:format
    return
  endif
  
  " Execute export command
  let cmd = g:lyra_repl_command . ' --export ' . a:format . ' --input "' . filename . '" --output "' . output_file . '"'
  let result = system(cmd)
  
  if v:shell_error == 0
    echo "Exported to " . a:format . ": " . output_file
  else
    echo "Export failed: " . result
  endif
endfunction

" Get function documentation
function! lyra#get_doc(word)
  let docs = {
        \ 'Sin': 'Sin[x] - Computes the sine of x (x in radians)',
        \ 'Cos': 'Cos[x] - Computes the cosine of x (x in radians)',
        \ 'Tan': 'Tan[x] - Computes the tangent of x (x in radians)',
        \ 'Log': 'Log[x] - Computes the natural logarithm of x',
        \ 'Exp': 'Exp[x] - Computes the exponential function e^x',
        \ 'Sqrt': 'Sqrt[x] - Computes the square root of x',
        \ 'Abs': 'Abs[x] - Computes the absolute value of x',
        \ 'Length': 'Length[list] - Returns the number of elements in list',
        \ 'Map': 'Map[f, list] - Applies function f to each element of list',
        \ 'Select': 'Select[list, pred] - Returns elements for which pred is True',
        \ 'If': 'If[condition, trueExpr, falseExpr] - Conditional expression',
        \ 'Pi': 'Pi - The mathematical constant π ≈ 3.14159',
        \ 'E': 'E - The mathematical constant e ≈ 2.71828'
        \ }
  
  if has_key(docs, a:word)
    return docs[a:word]
  else
    return 'No documentation available for: ' . a:word
  endif
endfunction

" Show function documentation in preview window
function! lyra#show_doc(word)
  let doc = lyra#get_doc(a:word)
  
  " Open preview window
  pedit __LyraDoc__
  wincmd P
  
  " Set up the preview buffer
  setlocal buftype=nofile
  setlocal noswapfile
  setlocal modifiable
  
  " Clear and add documentation
  silent %delete
  call append(0, [
        \ 'Lyra Documentation: ' . a:word,
        \ repeat('=', len('Lyra Documentation: ' . a:word)),
        \ '',
        \ doc
        \ ])
  
  " Remove the last empty line
  silent $delete
  
  " Make buffer read-only
  setlocal nomodifiable
  setlocal filetype=text
  
  " Return to original window
  wincmd p
endfunction

" Initialize completion for Lyra files
function! lyra#init_completion()
  setlocal omnifunc=lyra#complete
  setlocal completefunc=lyra#complete
endfunction