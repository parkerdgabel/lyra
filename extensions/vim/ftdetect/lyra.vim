" Vim filetype detection for Lyra
" Language: Lyra
" Maintainer: Lyra Language Team

" Detect Lyra files by extension
autocmd BufNewFile,BufRead *.lyra set filetype=lyra

" Detect Lyra files by shebang
autocmd BufNewFile,BufRead * if getline(1) =~ '^#!.*\blyra\b' | set filetype=lyra | endif

" Detect Lyra files by content patterns
autocmd BufNewFile,BufRead * if getline(1) =~ '^\s*%' && getline(2) =~ 'Lyra' | set filetype=lyra | endif

" Set default file type for new .lyra files
autocmd BufNewFile *.lyra 0put ='// Lyra mathematical computing script' | set filetype=lyra