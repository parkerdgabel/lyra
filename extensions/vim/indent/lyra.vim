" Vim indent file
" Language: Lyra
" Maintainer: Lyra Language Team

if exists("b:did_indent")
  finish
endif
let b:did_indent = 1

setlocal autoindent
setlocal smartindent
setlocal expandtab
setlocal tabstop=2
setlocal shiftwidth=2
setlocal softtabstop=2

" Enable automatic indentation
setlocal indentexpr=GetLyraIndent(v:lnum)
setlocal indentkeys=!^F,o,O,e,},],),=Module,=Block,=If,=Which,=For,=While

" Only define the function once
if exists("*GetLyraIndent")
  finish
endif

function! GetLyraIndent(lnum)
  " Find a non-blank line above the current line
  let prevlnum = prevnonblank(a:lnum - 1)
  
  " If there's no previous line, don't indent
  if prevlnum == 0
    return 0
  endif
  
  " Get the previous line
  let prevline = getline(prevlnum)
  let curline = getline(a:lnum)
  
  " Start with the indent of the previous line
  let indent = indent(prevlnum)
  
  " Increase indent after opening brackets or control structures
  if prevline =~ '\v[\[{(]\s*$'
    let indent += &shiftwidth
  endif
  
  " Increase indent after control structure keywords
  if prevline =~ '\v<(If|Which|For|While|Module|Block|Function)\s*\['
    let indent += &shiftwidth
  endif
  
  " Increase indent after assignment with multiline expressions
  if prevline =~ '\v:?\=\s*$' || prevline =~ '\v:?\=.*[\[{(]\s*$'
    let indent += &shiftwidth
  endif
  
  " Increase indent after rule operators
  if prevline =~ '\v(-\>|:\>)\s*$'
    let indent += &shiftwidth
  endif
  
  " Decrease indent for closing brackets
  if curline =~ '\v^\s*[\]})]'
    let indent -= &shiftwidth
  endif
  
  " Decrease indent for 'else' and similar keywords
  if curline =~ '\v^\s*(Else|ElseIf)\>'
    let indent -= &shiftwidth
  endif
  
  " Special handling for list elements
  if prevline =~ '\v,\s*$' && curline !~ '\v^\s*[\]})]'
    " Keep same indent as previous line for continued list elements
    return indent(prevlnum)
  endif
  
  " Special handling for function parameters
  if prevline =~ '\v\[\s*$' && curline !~ '\v^\s*\]'
    let indent += &shiftwidth
  endif
  
  " Ensure we don't go negative
  if indent < 0
    let indent = 0
  endif
  
  return indent
endfunction

" Set up bracket matching
setlocal matchpairs+=<:>

" Configure comment formatting
setlocal formatoptions-=t
setlocal formatoptions+=croql

" Set comment strings
setlocal commentstring=//\ %s
setlocal comments=://,s:/*,mb:*,ex:*/