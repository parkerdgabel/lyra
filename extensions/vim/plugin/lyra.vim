" Vim plugin for Lyra language support
" Language: Lyra
" Maintainer: Lyra Language Team
" Version: 1.0

if exists("g:loaded_lyra") || &cp
  finish
endif
let g:loaded_lyra = 1

" Default configuration
if !exists('g:lyra_repl_command')
  let g:lyra_repl_command = 'lyra'
endif

if !exists('g:lyra_repl_args')
  let g:lyra_repl_args = []
endif

if !exists('g:lyra_auto_start_repl')
  let g:lyra_auto_start_repl = 0
endif

if !exists('g:lyra_split_direction')
  let g:lyra_split_direction = 'vertical'
endif

if !exists('g:lyra_repl_size')
  let g:lyra_repl_size = 40
endif

" REPL management variables
let s:lyra_repl_job = 0
let s:lyra_repl_buffer = 0
let s:lyra_repl_window = 0

" Function to start Lyra REPL
function! s:StartLyraREPL()
  if s:lyra_repl_job != 0 && job_status(s:lyra_repl_job) == 'run'
    echo "Lyra REPL is already running"
    call s:ShowREPLWindow()
    return
  endif
  
  " Create new buffer for REPL
  if g:lyra_split_direction == 'horizontal'
    execute 'split'
    execute 'resize ' . g:lyra_repl_size
  else
    execute 'vsplit'
    execute 'vertical resize ' . g:lyra_repl_size
  endif
  
  " Create terminal with Lyra REPL
  let l:cmd = [g:lyra_repl_command] + g:lyra_repl_args
  
  if has('nvim')
    let s:lyra_repl_job = termopen(l:cmd)
    let s:lyra_repl_buffer = bufnr('%')
  else
    let s:lyra_repl_job = term_start(l:cmd, {
          \ 'term_name': 'Lyra REPL',
          \ 'vertical': g:lyra_split_direction == 'vertical',
          \ 'term_finish': 'close'
          \ })
    let s:lyra_repl_buffer = term_getbuf(s:lyra_repl_job)
  endif
  
  let s:lyra_repl_window = win_getid()
  
  " Set buffer options
  setlocal buftype=terminal
  setlocal noswapfile
  setlocal nomodified
  
  echo "Lyra REPL started"
endfunction

" Function to stop Lyra REPL
function! s:StopLyraREPL()
  if s:lyra_repl_job == 0
    echo "No Lyra REPL is running"
    return
  endif
  
  if has('nvim')
    call jobstop(s:lyra_repl_job)
  else
    call job_stop(s:lyra_repl_job)
  endif
  
  " Close REPL window if it exists
  if s:lyra_repl_window != 0 && win_id2win(s:lyra_repl_window) != 0
    call win_gotoid(s:lyra_repl_window)
    close
  endif
  
  let s:lyra_repl_job = 0
  let s:lyra_repl_buffer = 0
  let s:lyra_repl_window = 0
  
  echo "Lyra REPL stopped"
endfunction

" Function to restart Lyra REPL
function! s:RestartLyraREPL()
  call s:StopLyraREPL()
  sleep 500m
  call s:StartLyraREPL()
endfunction

" Function to show REPL window
function! s:ShowREPLWindow()
  if s:lyra_repl_buffer == 0
    echo "No Lyra REPL buffer available"
    return
  endif
  
  " Find if REPL window is already visible
  let l:winid = bufwinid(s:lyra_repl_buffer)
  if l:winid != -1
    call win_gotoid(l:winid)
    return
  endif
  
  " Create new window for existing REPL buffer
  if g:lyra_split_direction == 'horizontal'
    execute 'split'
    execute 'resize ' . g:lyra_repl_size
  else
    execute 'vsplit'
    execute 'vertical resize ' . g:lyra_repl_size
  endif
  
  execute 'buffer ' . s:lyra_repl_buffer
  let s:lyra_repl_window = win_getid()
endfunction

" Function to send text to REPL
function! s:SendToREPL(text)
  if s:lyra_repl_job == 0 || (has('nvim') ? 0 : job_status(s:lyra_repl_job) != 'run')
    echo "No running Lyra REPL found. Starting new one..."
    call s:StartLyraREPL()
    sleep 1000m  " Wait for REPL to start
  endif
  
  " Ensure REPL window is visible
  call s:ShowREPLWindow()
  
  " Send text to terminal
  if has('nvim')
    call chansend(s:lyra_repl_job, a:text . "\n")
  else
    call term_sendkeys(s:lyra_repl_buffer, a:text . "\n")
  endif
endfunction

" Function to evaluate current line
function! s:EvaluateLine()
  let l:line = getline('.')
  if l:line =~ '^\s*$'
    echo "Empty line, nothing to evaluate"
    return
  endif
  
  call s:SendToREPL(l:line)
  echo "Sent line to REPL: " . l:line
endfunction

" Function to evaluate visual selection
function! s:EvaluateSelection() range
  let l:lines = getline(a:firstline, a:lastline)
  let l:text = join(l:lines, "\n")
  
  if l:text =~ '^\s*$'
    echo "Empty selection, nothing to evaluate"
    return
  endif
  
  call s:SendToREPL(l:text)
  echo "Sent selection to REPL (" . len(l:lines) . " lines)"
endfunction

" Function to evaluate entire buffer
function! s:EvaluateBuffer()
  let l:lines = getline(1, '$')
  let l:text = join(l:lines, "\n")
  
  if l:text =~ '^\s*$'
    echo "Empty buffer, nothing to evaluate"
    return
  endif
  
  call s:SendToREPL(l:text)
  echo "Sent entire buffer to REPL (" . len(l:lines) . " lines)"
endfunction

" Function to load current file in REPL
function! s:LoadFile()
  if &filetype != 'lyra'
    echo "Current buffer is not a Lyra file"
    return
  endif
  
  let l:filename = expand('%')
  if l:filename == ''
    echo "Buffer has no associated file"
    return
  endif
  
  " Save file first if modified
  if &modified
    write
  endif
  
  call s:SendToREPL('%load "' . l:filename . '"')
  echo "Loading file in REPL: " . l:filename
endfunction

" Function to get help for word under cursor
function! s:GetHelp()
  let l:word = expand('<cword>')
  if l:word == ''
    echo "No word under cursor"
    return
  endif
  
  call s:SendToREPL('%help ' . l:word)
  echo "Getting help for: " . l:word
endfunction

" Commands
command! LyraREPLStart call s:StartLyraREPL()
command! LyraREPLStop call s:StopLyraREPL()
command! LyraREPLRestart call s:RestartLyraREPL()
command! LyraREPLShow call s:ShowREPLWindow()
command! LyraEvaluateLine call s:EvaluateLine()
command! -range LyraEvaluateSelection <line1>,<line2>call s:EvaluateSelection()
command! LyraEvaluateBuffer call s:EvaluateBuffer()
command! LyraLoadFile call s:LoadFile()
command! LyraHelp call s:GetHelp()

" Key mappings for Lyra files
augroup LyraKeyMappings
  autocmd!
  autocmd FileType lyra nnoremap <buffer> <leader>rs :LyraREPLStart<CR>
  autocmd FileType lyra nnoremap <buffer> <leader>rq :LyraREPLStop<CR>
  autocmd FileType lyra nnoremap <buffer> <leader>rr :LyraREPLRestart<CR>
  autocmd FileType lyra nnoremap <buffer> <leader>el :LyraEvaluateLine<CR>
  autocmd FileType lyra vnoremap <buffer> <leader>es :LyraEvaluateSelection<CR>
  autocmd FileType lyra nnoremap <buffer> <leader>eb :LyraEvaluateBuffer<CR>
  autocmd FileType lyra nnoremap <buffer> <leader>lf :LyraLoadFile<CR>
  autocmd FileType lyra nnoremap <buffer> K :LyraHelp<CR>
  autocmd FileType lyra nnoremap <buffer> <F5> :LyraEvaluateBuffer<CR>
  autocmd FileType lyra inoremap <buffer> <F5> <Esc>:LyraEvaluateBuffer<CR>
augroup END

" Auto-start REPL if configured
augroup LyraAutoStart
  autocmd!
  if g:lyra_auto_start_repl
    autocmd FileType lyra call s:StartLyraREPL()
  endif
augroup END

" Folding support
augroup LyraFolding
  autocmd!
  autocmd FileType lyra setlocal foldmethod=syntax
  autocmd FileType lyra setlocal foldlevel=99
augroup END

" Set up menu items (for GUI Vim)
if has("gui_running")
  amenu &Lyra.Start\ REPL :LyraREPLStart<CR>
  amenu &Lyra.Stop\ REPL :LyraREPLStop<CR>
  amenu &Lyra.Restart\ REPL :LyraREPLRestart<CR>
  amenu &Lyra.-sep1- :
  amenu &Lyra.Evaluate\ Line :LyraEvaluateLine<CR>
  amenu &Lyra.Evaluate\ Selection :LyraEvaluateSelection<CR>
  amenu &Lyra.Evaluate\ Buffer :LyraEvaluateBuffer<CR>
  amenu &Lyra.-sep2- :
  amenu &Lyra.Load\ File :LyraLoadFile<CR>
  amenu &Lyra.Help\ for\ Word :LyraHelp<CR>
endif

" Status line integration
function! LyraREPLStatus()
  if s:lyra_repl_job != 0
    if has('nvim') || job_status(s:lyra_repl_job) == 'run'
      return '[REPL:ON]'
    else
      return '[REPL:OFF]'
    endif
  else
    return '[REPL:OFF]'
  endif
endfunction

" Add to status line for Lyra files
augroup LyraStatusLine
  autocmd!
  autocmd FileType lyra setlocal statusline+=%{LyraREPLStatus()}
augroup END