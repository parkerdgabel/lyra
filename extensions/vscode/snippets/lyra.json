{
  "Function Definition": {
    "prefix": "func",
    "body": [
      "${1:functionName}[${2:args}_] := ${3:body}"
    ],
    "description": "Define a function with delayed evaluation"
  },
  "Immediate Function Definition": {
    "prefix": "funcimm",
    "body": [
      "${1:functionName}[${2:args}_] = ${3:body}"
    ],
    "description": "Define a function with immediate evaluation"
  },
  "If Statement": {
    "prefix": "if",
    "body": [
      "If[${1:condition}, ${2:trueResult}, ${3:falseResult}]"
    ],
    "description": "Conditional expression"
  },
  "Which Statement": {
    "prefix": "which",
    "body": [
      "Which[",
      "  ${1:condition1}, ${2:result1},",
      "  ${3:condition2}, ${4:result2},",
      "  True, ${5:defaultResult}",
      "]"
    ],
    "description": "Multiple condition statement"
  },
  "For Loop": {
    "prefix": "for",
    "body": [
      "For[${1:i} = ${2:start}, ${1:i} <= ${3:end}, ${1:i}++,",
      "  ${4:body}",
      "]"
    ],
    "description": "For loop"
  },
  "While Loop": {
    "prefix": "while",
    "body": [
      "While[${1:condition},",
      "  ${2:body}",
      "]"
    ],
    "description": "While loop"
  },
  "Module": {
    "prefix": "module",
    "body": [
      "Module[{${1:localVars}},",
      "  ${2:body}",
      "]"
    ],
    "description": "Module with local variables"
  },
  "Block": {
    "prefix": "block",
    "body": [
      "Block[{${1:localVars}},",
      "  ${2:body}",
      "]"
    ],
    "description": "Block with local variables"
  },
  "List": {
    "prefix": "list",
    "body": [
      "{${1:element1}, ${2:element2}, ${3:element3}}"
    ],
    "description": "Create a list"
  },
  "Rule": {
    "prefix": "rule",
    "body": [
      "${1:pattern} -> ${2:replacement}"
    ],
    "description": "Replacement rule"
  },
  "Delayed Rule": {
    "prefix": "drule",
    "body": [
      "${1:pattern} :> ${2:replacement}"
    ],
    "description": "Delayed replacement rule"
  },
  "Replace All": {
    "prefix": "replaceall",
    "body": [
      "${1:expression} /. ${2:rules}"
    ],
    "description": "Replace all occurrences"
  },
  "Pattern": {
    "prefix": "pattern",
    "body": [
      "${1:name}_${2:type}"
    ],
    "description": "Named pattern with type"
  },
  "Blank Sequence": {
    "prefix": "blankseq",
    "body": [
      "${1:name}__${2:type}"
    ],
    "description": "Blank sequence pattern"
  },
  "Blank Null Sequence": {
    "prefix": "blanknull",
    "body": [
      "${1:name}___${2:type}"
    ],
    "description": "Blank null sequence pattern"
  },
  "Map": {
    "prefix": "map",
    "body": [
      "Map[${1:function}, ${2:list}]"
    ],
    "description": "Map function over list"
  },
  "Apply": {
    "prefix": "apply",
    "body": [
      "Apply[${1:function}, ${2:arguments}]"
    ],
    "description": "Apply function to arguments"
  },
  "Select": {
    "prefix": "select",
    "body": [
      "Select[${1:list}, ${2:predicate}]"
    ],
    "description": "Select elements matching predicate"
  },
  "Cases": {
    "prefix": "cases",
    "body": [
      "Cases[${1:expression}, ${2:pattern}]"
    ],
    "description": "Extract cases matching pattern"
  },
  "Table": {
    "prefix": "table",
    "body": [
      "Table[${1:expression}, {${2:iterator}, ${3:start}, ${4:end}}]"
    ],
    "description": "Generate table of values"
  },
  "Sum": {
    "prefix": "sum",
    "body": [
      "Sum[${1:expression}, {${2:iterator}, ${3:start}, ${4:end}}]"
    ],
    "description": "Symbolic sum"
  },
  "Product": {
    "prefix": "product",
    "body": [
      "Product[${1:expression}, {${2:iterator}, ${3:start}, ${4:end}}]"
    ],
    "description": "Symbolic product"
  },
  "Solve": {
    "prefix": "solve",
    "body": [
      "Solve[${1:equations}, ${2:variables}]"
    ],
    "description": "Solve equations"
  },
  "Plot": {
    "prefix": "plot",
    "body": [
      "Plot[${1:function}, {${2:variable}, ${3:min}, ${4:max}}]"
    ],
    "description": "Plot function"
  },
  "Matrix": {
    "prefix": "matrix",
    "body": [
      "{{${1:a11}, ${2:a12}}, {${3:a21}, ${4:a22}}}"
    ],
    "description": "2x2 matrix"
  },
  "Try Catch": {
    "prefix": "try",
    "body": [
      "TryCatch[",
      "  ${1:expression},",
      "  ${2:catchExpression}",
      "]"
    ],
    "description": "Try-catch error handling"
  },
  "Print Statement": {
    "prefix": "print",
    "body": [
      "Print[${1:message}]"
    ],
    "description": "Print message"
  },
  "Echo Statement": {
    "prefix": "echo",
    "body": [
      "Echo[${1:expression}]"
    ],
    "description": "Echo expression and return it"
  }
}