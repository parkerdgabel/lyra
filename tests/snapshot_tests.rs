use insta::assert_yaml_snapshot;
use lyra::{compiler::Compiler, parser::Parser};

/// Helper function to evaluate an expression and return the formatted result
fn evaluate_expression(source: &str) -> String {
    let mut parser = Parser::from_source(source).unwrap();
    let statements = parser.parse().unwrap();
    let expr = &statements[0];

    match Compiler::eval(expr) {
        Ok(value) => format_value(&value),
        Err(e) => format!("Error: {}", e),
    }
}

/// Format a value for consistent output
fn format_value(value: &lyra::vm::Value) -> String {
    match value {
        lyra::vm::Value::Integer(n) => n.to_string(),
        lyra::vm::Value::Real(f) => {
            if f.fract() == 0.0 {
                format!("{:.1}", f)
            } else {
                format!("{:.6}", f) // Fixed precision for consistent snapshots
            }
        }
        lyra::vm::Value::String(s) => format!("\"{}\"", s),
        lyra::vm::Value::Symbol(s) => s.clone(),
        lyra::vm::Value::List(items) => {
            let formatted_items: Vec<String> = items.iter().map(format_value).collect();
            format!("{{{}}}", formatted_items.join(", "))
        }
        lyra::vm::Value::Function(name) => format!("Function[{}]", name),
        lyra::vm::Value::Boolean(b) => if *b { "True" } else { "False" }.to_string(),
        lyra::vm::Value::Tensor(tensor) => {
            format!("Tensor[shape: {:?}, elements: {}]", 
                    tensor.shape(), 
                    tensor.len())
        }
        lyra::vm::Value::Missing => "Missing[]".to_string(),
        // Note: Series, Table, Dataset, Schema are now handled by LyObj case below
        lyra::vm::Value::LyObj(obj) => {
            format!("{}[...]", obj.type_name())
        }
    }
}

#[test]
fn test_arithmetic_expressions_snapshot() {
    let expressions = vec![
        "2 + 3",
        "10 - 4",
        "3 * 7",
        "15 / 3",
        "2^8",
        "2 + 3 * 4",
        "(2 + 3) * 4",
        "2^3 * 4 + 1",
        "2^(3 + 1)",
        "(5 + 3) / (2 * 2)",
    ];

    let results: Vec<String> = expressions
        .iter()
        .map(|expr| evaluate_expression(expr))
        .collect();

    assert_yaml_snapshot!("arithmetic_expressions", results);
}

#[test]
fn test_mathematical_functions_snapshot() {
    let expressions = vec![
        "Sin[0]", "Cos[0]", "Tan[0]", "Exp[0]", "Exp[1]", "Log[1]", "Sqrt[4]", "Sqrt[16]", "2^10",
        "10^3",
    ];

    let results: Vec<String> = expressions
        .iter()
        .map(|expr| evaluate_expression(expr))
        .collect();

    assert_yaml_snapshot!("mathematical_functions", results);
}

#[test]
fn test_list_operations_snapshot() {
    let expressions = vec![
        "{1, 2, 3, 4, 5}",
        "Length[{1, 2, 3, 4, 5}]",
        "Head[{10, 20, 30}]",
        "Tail[{10, 20, 30}]",
        "Append[{1, 2, 3}, 4]",
        "Length[{}]",
        "Append[{}, \"first\"]",
        "{\"apple\", \"banana\", \"cherry\"}",
    ];

    let results: Vec<String> = expressions
        .iter()
        .map(|expr| evaluate_expression(expr))
        .collect();

    assert_yaml_snapshot!("list_operations", results);
}

#[test]
fn test_string_operations_snapshot() {
    let expressions = vec![
        "\"Hello, World!\"",
        "StringLength[\"Hello\"]",
        "StringLength[\"\"]",
        "StringJoin[\"Hello\", \" \", \"World\"]",
        "StringTake[\"Programming\", 4]",
        "StringDrop[\"Programming\", 4]",
        "StringLength[StringJoin[\"Hello\", \" \", \"World\"]]",
    ];

    let results: Vec<String> = expressions
        .iter()
        .map(|expr| evaluate_expression(expr))
        .collect();

    assert_yaml_snapshot!("string_operations", results);
}

#[test]
fn test_complex_expressions_snapshot() {
    let expressions = vec![
        "Sin[3.14159 / 2] + Cos[0]",
        "Sqrt[3^2 + 4^2]",
        "Length[Append[{1, 2, 3}, Sqrt[16]]]",
        "StringLength[StringJoin[\"Result: \", \"42\"]]",
        "((2 + 3) * (4 + 5)) / ((6 + 7) - (8 - 9))",
    ];

    let results: Vec<String> = expressions
        .iter()
        .map(|expr| evaluate_expression(expr))
        .collect();

    assert_yaml_snapshot!("complex_expressions", results);
}

#[test]
fn test_edge_cases_snapshot() {
    let expressions = vec![
        "0 + 0",
        "1 * 1",
        "0 / 1",
        "Sqrt[0]",
        "Log[1]",
        "Sin[0]",
        "Cos[0]",
        "Exp[0]",
        "Length[{}]",
        "StringLength[\"\"]",
    ];

    let results: Vec<String> = expressions
        .iter()
        .map(|expr| evaluate_expression(expr))
        .collect();

    assert_yaml_snapshot!("edge_cases", results);
}
