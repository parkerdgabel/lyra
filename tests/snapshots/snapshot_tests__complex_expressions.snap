---
source: tests/snapshot_tests.rs
expression: results
---
- "2.000000"
- "5.0"
- "4"
- "10"
- "3.214286"
